internal/k8s/controller.go

```diff
@@ -1551,32 +1566,37 @@ func (lbc *LoadBalancerController) createVirtualServer(virtualServer *conf_v1alp
 		virtualServerRoutes = append(virtualServerRoutes, vsr)
 
 		for _, u := range vsr.Spec.Upstreams {
-			endpointsKey := configs.GenerateEndpointsKey(vsr.Namespace, u.Service, u.Port)
-			endpoints[endpointsKey] = lbc.getEndpointsForService(vsr.Namespace, u.Service, int(u.Port))
+			backend := &extensions.IngressBackend{
+				ServiceName: u.Service,
+				ServicePort: intstr.FromInt(int(u.Port)),
+			}
+
+			svc, err := lbc.getServiceForIngressBackend(backend, vsr.Namespace)
+			if err != nil {
+				glog.V(3).Infof("Error getting service %v: %v", u.Service, err)
+			} else {
+				endpointsKey := configs.GenerateEndpointsKey(vsr.Namespace, u.Service, u.Port)
+				endps, external, err := lbc.getEndpointsForIngressBackend(backend, vsr.Namespace, svc)
+				if err == nil && external && lbc.isNginxPlus {
+					externalNameSvcs[u.Service] = true
+				}
+				endpoints[endpointsKey] = endps
+			}
 		}
 	}
 
 	virtualServerEx.Endpoints = endpoints
 	virtualServerEx.VirtualServerRoutes = virtualServerRoutes
+	virtualServerEx.ExternalNameSvcs = externalNameSvcs
 
 	return &virtualServerEx, virtualServerRouteErrors
 }
 
-func (lbc *LoadBalancerController) getEndpointsForService(namespace string, name string, port int) []string {
```

Comment:
please note this function existed to separate any Ingress-related code (functions like `getServiceForIngressBackend` and structs like `extensions.IngressBackend` from the virtual server related code so that in `createVirtualServer` we don't use any Ingress related code directly.

Reply:
That's what I thought at the beginning, but then I saw that `getServiceForIngressBackend` was called inside, so it's not really separated, is just a useless abstraction. Better to split in a more clever way I would say.